src/Microsoft.ML.Recommender/MatrixFactorizationPredictor.cs

```diff
@@ -43,33 +46,51 @@ private static VersionInfo GetVersionInfo()
+                loaderAssemblyName: typeof(MatrixFactorizationModelParameters).Assembly.FullName);
         }
         private const uint VersionNoMinCount = 0x00010002;
 
         private readonly IHost _host;
-        // The number of rows.
-        private readonly int _numberOfRows;
-        // The number of columns.
-        private readonly int _numberofColumns;
-        // The rank of the factor matrices.
-        private readonly int _approximationRank;
-        // Packed _numberOfRows by _approximationRank matrix.
+        ///<summary> The number of rows.</summary>
+        public readonly int NumberOfRows;
+        ///<summary> The number of columns.</summary>
+        public readonly int NumberOfColumns;
+        ///<summary> The rank of the factor matrices.</summary>
+        public readonly int ApproximationRank;
+        // Packed NumberOfRows by ApproximationRank matrix.
         private readonly float[] _leftFactorMatrix;
-        // Packed _approximationRank by _numberofColumns matrix.
+        // Packed ApproximationRank by NumberOfColumns matrix.
         private readonly float[] _rightFactorMatrix;
 
-        public PredictionKind PredictionKind
+        /// <summary>
+        /// Returns left approximation matrix.
+        /// </summary>
+        /// <returns>Flattened into one-dimensional array with size equal to <see cref="NumberOfRows"/> * <see cref="ApproximationRank"/></returns>
+        public float[] GetLeftFactorMatrix()
         {
-            get { return PredictionKind.Recommendation; }
+            float[] result = new float[NumberOfRows * ApproximationRank];
+            _leftFactorMatrix.CopyTo(result, 0);
+            return result;
         }
+        /// <summary>
+        /// Returns right approximation matrix.
+        /// </summary>
+        /// <returns>Flattened into one-dimensional array with size equal to <see cref="ApproximationRank"/> * <see cref="NumberOfColumns"/></returns>
+        public float[] GetRightFactorMatrix()
+        {
+            float[] result = new float[ApproximationRank* NumberOfColumns];
+            _rightFactorMatrix.CopyTo(result, 0);
+            return result;
+        }
+
+        public PredictionKind PredictionKind => PredictionKind.Recommendation;
 
-        public ColumnType OutputType { get { return NumberType.Float; } }
+        private ColumnType OutputType { get { return NumberType.Float; } }
 
-        public ColumnType MatrixColumnIndexType { get; }
-        public ColumnType MatrixRowIndexType { get; }
+        internal ColumnType MatrixColumnIndexType { get; }
+        internal ColumnType MatrixRowIndexType { get; }
 
-        internal MatrixFactorizationPredictor(IHostEnvironment env, SafeTrainingAndModelBuffer buffer, KeyType matrixColumnIndexType, KeyType matrixRowIndexType)
+        internal MatrixFactorizationModelParameters(IHostEnvironment env, SafeTrainingAndModelBuffer buffer, KeyType matrixColumnIndexType, KeyType matrixRowIndexType)
         {
```

Comment:
Do we have a public constructor for creating a MatrixFactorizationModelParameters from its public fields?

Reply:
No we don't.
See [https://github.com/dotnet/machinelearning/issues/1866](https://github.com/dotnet/machinelearning/issues/1866)

---
In reply to: [254909189](https://github.com/dotnet/machinelearning/pull/2451#discussion_r254909189) [](ancestors = 254909189)